### ISSUE TITLE
XML Deserialization in Microsoft.Extensions.Configuration - Parity with Data Contracts

### ISSUE THREAD

**Issue by maryamariyan** [dotnet/runtime#64840]
Based on docs for **Interchangeable collections** in [Collection Types in Data Contracts](https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/collection-types-in-data-contracts#interchangeable-collections), we see that:

```c#
[DataContract(Name = "PurchaseOrder")]
public class PurchaseOrder1
{
    [DataMember]
    public string customerName;
    [DataMember]
    public Collection<Item> items;
    [DataMember]
    public string[] comments;
}
```
and 
```c#
[DataContract(Name = "PurchaseOrder")]
public class PurchaseOrder2
{
    [DataMember]
    public string customerName;
    [DataMember]
    public List<Item> items;
    [DataMember]
    public BindingList<string> comments;
}
```
both get an XML data contract similar to the following code:

```xml
<PurchaseOrder>
    <customerName>...</customerName>
    <items>
        <Item>...</Item>
        <Item>...</Item>
        <Item>...</Item>
        ...
    </items>
    <comments>
        <string>...</string>
        <string>...</string>
        <string>...</string>
        ...
    </comments>
</PurchaseOrder>
```

However, using `Microsoft.Extensions.Configuration.Xml` to deserialize to XML we require a wrapping class like `ItemList` similar to the format below:

```c#
class PurchaseOrder {
                public string CustomerName {get; set;}
                public ItemList Items {get;set;}
}
class ItemList {
                public Item[] Item
} 
class Item {…}
```

To help with developers trying to migrate from Data Contract to `Microsoft.Extensions.Configuration`, we are looking to update the extensions library to deserialize without the need of `ItemList`, like below:

```c#
[DataContract]
class PurchaseOrder {
                [DataMember]
                public string CustomerName {get; set;}
                [DataMember] 
                public Item[] Items {get;set;}
}
[DataContract]
class Item {…}
```

## API Proposal

With the new behaviour of XML deserialization, this wrapping class `ItemList` would no longer be needed.

To accomplish this :

1 - We add a new configuration flag called `IgnoreElementNameForRepeats` this way we avoid introducing a breaking change to existing deserialization behaviour.
2 - On unexpected xml input for this special migration use case, when the configuration flag is enabled we throw exception.

```diff
    public partial class XmlConfigurationSource : Microsoft.Extensions.Configuration.FileConfigurationSource
    {
        public XmlConfigurationSource() { }
+        public bool IgnoreElementNameForRepeats { get; set; }
        public override Microsoft.Extensions.Configuration.IConfigurationProvider Build(Microsoft.Extensions.Configuration.IConfigurationBuilder builder) { throw null; }
    }
    public partial class XmlStreamConfigurationProvider : Microsoft.Extensions.Configuration.StreamConfigurationProvider
    {
        public XmlStreamConfigurationProvider(Microsoft.Extensions.Configuration.Xml.XmlStreamConfigurationSource source) : base (default(Microsoft.Extensions.Configuration.StreamConfigurationSource)) { }
        public override void Load(System.IO.Stream stream) { }
        public static System.Collections.Generic.IDictionary<string, string> Read(System.IO.Stream stream, Microsoft.Extensions.Configuration.Xml.XmlDocumentDecryptor decryptor) { throw null; }
+        public static System.Collections.Generic.IDictionary<string, string> Read(System.IO.Stream stream, Microsoft.Extensions.Configuration.Xml.XmlDocumentDecryptor decryptor, bool ignoreElementNameForRepeats) { throw null; }
    }
    public partial class XmlStreamConfigurationSource : Microsoft.Extensions.Configuration.StreamConfigurationSource
    {
        public XmlStreamConfigurationSource() { }
+        public bool IgnoreElementNameForRepeats { get; set; }
        public override Microsoft.Extensions.Configuration.IConfigurationProvider Build(Microsoft.Extensions.Configuration.IConfigurationBuilder builder) { throw null; }
    }
```

Refer to drafted prototype PR by @vidommet: https://github.com/dotnet/runtime/pull/64686

## Risk
Low. Since with this solution we are adding a configuration flag, disabled by default, we would not introduce a breaking change to the existing deserialization behaviour.

**Comment by maryamariyan** [comment]
@vidommet could you please elaborate more on this migration scenario? 

Is there a good indication for us to know if this kind of problem is something many developers wanting to use extension's xml configuration library today might find this feature useful? 

cc @amoerie seems like this feature new addition is in the same area as the feature you included last year in https://github.com/dotnet/runtime/pull/44608

**Comment by amoerie** [comment]
I can't speak for everyone, but our use case was mixing JSON and XML configuration files, mostly because our installer did not support JSON files out of the box.

This meant that the produced keys had to be similar to what the JSON configuration provider produces. This was a big focus of my PR, which tried to consolidate how parsing arrays worked between XML and JSON.

Here, the author seems to be looking for symmetry between generating XML from C# classes and subsequently using that as a configuration source. I can understand the idea, but we never generated our config files programmatically, so this was not relevant for our use case.

I also wonder: if you're generating your config files anyway, why use XML at all?

**Comment by vidommet** [comment]
Our current situation is do migrate from existing `DataContractSerializer` to `Extensions.Configuration.Xml`. Most of the XML in this situation is hand-crafted for easy readability. This feature makes it straightforward from existing [format](https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/collection-types-in-data-contracts#interchangeable-collections) rather than modifying each of the XML files one by one to the format required by `Configuration.Xml`.

Also, this feature improves the readability of XML and its corresponding POCO configuration classes. Simple items with arrays woudn't need an enclosing array as written in the proposal.

The situation is similar with teams trying to migrate from existing Data contract serialization to the extensions libraries.

**Comment by amoerie** [comment]
@vidommet I'm still trying to figure out what you mean by migration. 
Extensions.Configuration.Xml is just another provider in the Extensions.Configuration ecosystem that produces a list of key values for a given XML file. 
DataContractSerializer on the other hand is a serialization / deserialization mechanism that can turn C# POCOs into XML and vice versa.
The example XML input files that you give are perfectly supported by the existing implementation of the XML provider in .NET 6. So should I assume that the produced keys are not what you expect/want? 

Your example file

```xml
<PurchaseOrder>
    <customerName>...</customerName>
    <items>
        <Item>...</Item>
        <Item>...</Item>
        <Item>...</Item>
        ...
    </items>
    <comments>
        <string>...</string>
        <string>...</string>
        <string>...</string>
        ...
    </comments>
</PurchaseOrder>
```

Would produce the following key values (IIRC):

```
customerName = ..
items:item:0 = ...
items:item:1 = ...
items:item:2 = ...
comments:string:0 = ...
comments:string:1 = ...
comments:string:2 = ...
```

**Comment by vidommet** [comment]
@amoerie that is correct. In the sample, you provided, `items:item:0` requires two classes for us to be able to bind to a POCO. Compare this to data contract serializer, which only needs one. This makes migration from one system to another much harder because we have to maintain two sets of POCOs.

The proposal here is to make the sample you provided into something like the following which will remove the need for two sets of POCOs.
```
customerName = ..
items:0 = ...
items:1 = ...
items:2 = ...
comments:0 = ...
comments:1 = ...
comments:2 = ..
```

**Comment by ericstj** [comment]
> Extensions.Configuration.Xml is just another provider in the Extensions.Configuration ecosystem that produces a list of key values for a given XML file.
DataContractSerializer on the other hand is a serialization / deserialization mechanism that can turn C# POCOs into XML and vice versa.

I agree with this 💯.  We'll never get parity with DCS -- that's a non-goal.  Just like it's a non-goal to get to parity with features of other serializers like System.Text.Json or Newtonsoft.Json.  Those are serializers.  This is a configuration API.

> This makes migration from one system to another

Can you tell us why you're trying to migrate from one to the other?  DataContractSerializer should work just fine on .NETCore and give you better performance for this sort of task than Configuration.

**Comment by vidommet** [comment]
DataContractSerializer forces us to use XML and doesn’t provide all the flexibilities of Extensions.Configuration including OOB support for non-XML formats. We are expanding the capabilities of our library to support a broad list of formats and validation options. Extensions.Configuration is a much better option with this when compared to DataContractSerializer.

I am not trying to bring feature parity with DCS. The aim is to provide interoperability of the POCOs so that switching between them is simpler. Currently POCOs accepted by both are similar with a few differences. I am trying to reduce them so that there is a path for the switch.

Performance is not a problem here because we don't load configurations for every request. We load them once and reuse the objects throughout the lifetime of the application until the next change in the configuration.

**Comment by ericstj** [comment]
The problem here is that configuration is never going to have the sort of features that serializers do.  Things that let folks be opinionated about the format of the input files or how conversions are done.  Its purpose is simpler.  Why should we add policy to the provider for this one case?  Others will certainly have more.

Can't the path be "convert your XML file by removing this node"?  If your goal is a one time move from DCS to some new solution that seems like a better tradeoff.  I imagine a tool could easily identify nodes to remove or even a do it at runtime if you didn't care much about the performance.  Lots of potential solutions.

**Comment by rosebyte** [comment]
Triage: we agree with @ericstj
